storm-core/src/jvm/org/apache/storm/StormSubmitter.java

```diff
@@ -238,39 +239,24 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
                 }
 
                 // Dependency uploading only makes sense for distributed mode
-                List<String> jarsBlobKeys;
+                List<String> jarsBlobKeys = Collections.emptyList();
                 List<String> artifactsBlobKeys;
 
                 DependencyUploader uploader = new DependencyUploader();
                 try {
                     uploader.init();
                     jarsBlobKeys = uploadDependencyJarsToBlobStore(uploader);
                     artifactsBlobKeys = uploadDependencyArtifactsToBlobStore(uploader);
+                    setDependencyBlobsToTopology(topology, jarsBlobKeys, artifactsBlobKeys);
+                    submitTopologyInDistributeMode(name, topology, opts, progressListener, asUser, conf, serConf);
+                } catch (Throwable e) {
+                    // remove uploaded jars blobs, not artifacts since they're shared across the cluster
+                    uploader.deleteBlobs(jarsBlobKeys);
```

Comment:
@HeartSaVioR Is not there a possibility that the topology is submitted successfully on nimbus but client receives thrift error because of some timeout or connection error? 
My preference is to remove those topologies only when we know definitely that the topology submission is failed.